service/controller/resource/deployment/securityrules.go

```diff
@@ -0,0 +1,409 @@
+package deployment
+
+import (
+	"context"
+	"strconv"
+
+	azurenetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-11-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+func (r Resource) getMasterSecurityRules(ctx context.Context, customObject providerv1alpha1.AzureConfig) ([]azurenetwork.SecurityRule, error) {
+	return r.getCustomSecurityRules(ctx, customObject, r.getDefaultMasterSecurityRules(customObject), key.MasterSecurityGroupName(customObject))
+}
+
+func (r Resource) getWorkersSecurityRules(ctx context.Context, customObject providerv1alpha1.AzureConfig) ([]azurenetwork.SecurityRule, error) {
+	return r.getCustomSecurityRules(ctx, customObject, r.getDefaultWorkersSecurityRules(customObject), key.WorkerSecurityGroupName(customObject))
+}
+
+func (r Resource) getCustomSecurityRules(ctx context.Context, customObject providerv1alpha1.AzureConfig, defaultRules []azurenetwork.SecurityRule, sgName string) ([]azurenetwork.SecurityRule, error) {
+	securityRulesClient, err := r.clientFactory.GetSecurityRulesClient(customObject.Spec.Azure.CredentialSecret.Namespace, customObject.Spec.Azure.CredentialSecret.Name)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	iterator, err := securityRulesClient.ListComplete(ctx, customObject.Name, sgName)
+	if IsNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "warning", "message", "SecurityGroup %s not found: unable to check for existing security rules.", sgName)
+		return defaultRules, nil
+	} else if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	for iterator.NotDone() {
+		rule := iterator.Value()
+
+		// Check if rule is a default one by comparing the name.
+		isDefault := false
+		{
+			for _, candidate := range defaultRules {
+				if *rule.Name == *candidate.Name {
```

Comment:
What if _something_ modifies a default security rule? With this solution the modification would be lost, the default rule would be used instead. Not sure if this could ever happen.

Reply:
That's true. The assumption here is that security rules are immutable. 
If you need to override a default security rule you need a new one with lower priority.
I guess we can't do much but accept this limitation